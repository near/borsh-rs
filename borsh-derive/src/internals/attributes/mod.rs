use syn::{Attribute, Path};

pub mod field;
pub mod item;
pub mod parsing;

#[derive(Copy, Clone, PartialEq, PartialOrd, Eq, Ord)]
pub struct Symbol {
    /// Name of the attribute
    pub name: &'static str,
    /// Expected value format representation for error printing
    pub expected: &'static str,
    support: AsyncSupport,
}

#[derive(Copy, Clone, PartialEq, PartialOrd, Eq, Ord)]
enum AsyncSupport {
    Sync,
    Async,
    Both,
}

impl Symbol {
    pub const fn new(name: &'static str, expected: &'static str) -> Self {
        Self {
            name,
            expected,
            support: AsyncSupport::Both,
        }
    }

    pub const fn new_sync(name: &'static str, expected: &'static str) -> Self {
        Self {
            name,
            expected,
            support: AsyncSupport::Sync,
        }
    }

    pub const fn new_async(name: &'static str, expected: &'static str) -> Self {
        Self {
            name,
            expected,
            support: AsyncSupport::Async,
        }
    }

    pub const fn test_support<const ASYNC_OR_SYNC_SWITCH: bool>(&self) -> bool {
        if ASYNC_OR_SYNC_SWITCH {
            matches!(self.support, AsyncSupport::Async | AsyncSupport::Both)
        } else {
            matches!(self.support, AsyncSupport::Sync | AsyncSupport::Both)
        }
    }
}

/// `borsh` - top level prefix in nested meta attribute
pub const BORSH: Symbol = Symbol::new("borsh", "borsh(...)");
/// `bound` - sub-borsh nested meta, field-level only; `BorshSerialize` and `BorshDeserialize` contexts
pub const BOUND: Symbol = Symbol::new_sync("bound", "bound(...)");
/// `async_bound` - sub-borsh nested meta, field-level only; `BorshSerializeAsync` and `BorshDeserializeAsync` contexts
#[cfg(feature = "async")]
pub const ASYNC_BOUND: Symbol = Symbol::new_async("async_bound", "async_bound(...)");
///  `use_discriminant` - sub-borsh nested meta, item-level only, enums only;
/// `BorshSerialize`, `BorshDeserialize`, `BorshSerializeAsync` and `BorshDeserializeAsync` contexts
pub const USE_DISCRIMINANT: Symbol = Symbol::new("use_discriminant", "use_discriminant = ...");
/// `serialize` - sub-bound nested meta attribute
pub const SERIALIZE: Symbol = Symbol::new("serialize", "serialize = ...");
/// `deserialize` - sub-bound nested meta attribute
pub const DESERIALIZE: Symbol = Symbol::new("deserialize", "deserialize = ...");
/// `skip` - sub-borsh nested meta, field-level only attribute;
/// `BorshSerialize`, `BorshDeserialize`, `BorshSerializeAsync`, `BorshDeserializeAsync` and `BorshSchema` contexts
pub const SKIP: Symbol = Symbol::new("skip", "skip");
/// `init` - sub-borsh nested meta, item-level only attribute; `BorshDeserialize` and `BorshDeserializeAsync` contexts
pub const INIT: Symbol = Symbol::new("init", "init = ...");
/// `serialize_with` - sub-borsh nested meta, field-level only; `BorshSerialize` context
pub const SERIALIZE_WITH: Symbol = Symbol::new_sync("serialize_with", "serialize_with = ...");
/// `deserialize_with` - sub-borsh nested meta, field-level only; `BorshDeserialize` context
pub const DESERIALIZE_WITH: Symbol = Symbol::new_sync("deserialize_with", "deserialize_with = ...");
/// `serialize_with_async` - sub-borsh nested meta, field-level only; `BorshSerializeAsync` context
#[cfg(feature = "async")]
pub const SERIALIZE_WITH_ASYNC: Symbol =
    Symbol::new_async("serialize_with_async", "serialize_with_async = ...");
/// `deserialize_with_async` - sub-borsh nested meta, field-level only; `BorshDeserializeAsync` context
#[cfg(feature = "async")]
pub const DESERIALIZE_WITH_ASYNC: Symbol =
    Symbol::new_async("deserialize_with_async", "deserialize_with_async = ...");
/// `crate` - sub-borsh nested meta, item-level only;
/// `BorshSerialize`, `BorshDeserialize`, `BorshSerializeAsync`, `BorshDeserializeAsync` and `BorshSchema` contexts
pub const CRATE: Symbol = Symbol::new("crate", "crate = ...");

#[cfg(feature = "schema")]
pub mod schema_keys {
    use super::Symbol;

    /// `schema` - sub-borsh nested meta, `BorshSchema` context
    pub const SCHEMA: Symbol = Symbol::new("schema", "schema(...)");
    /// `params` - sub-schema nested meta, field-level only attribute
    pub const PARAMS: Symbol = Symbol::new("params", "params = ...");
    /// `serialize_with` - sub-borsh nested meta, field-level only, `BorshSerialize` context
    /// `with_funcs` - sub-schema nested meta, field-level only attribute
    pub const WITH_FUNCS: Symbol = Symbol::new("with_funcs", "with_funcs(...)");
    /// `declaration` - sub-with_funcs nested meta, field-level only attribute
    pub const DECLARATION: Symbol = Symbol::new("declaration", "declaration = ...");
    /// `definitions` - sub-with_funcs nested meta, field-level only attribute
    pub const DEFINITIONS: Symbol = Symbol::new("definitions", "definitions = ...");
}

#[derive(Clone, Copy)]
pub enum BoundType {
    Serialize,
    Deserialize,
}
impl PartialEq<Symbol> for Path {
    fn eq(&self, word: &Symbol) -> bool {
        self.is_ident(word.name)
    }
}

impl<'a> PartialEq<Symbol> for &'a Path {
    fn eq(&self, word: &Symbol) -> bool {
        self.is_ident(word.name)
    }
}

fn get_one_attribute(attrs: &[Attribute]) -> syn::Result<Option<&Attribute>> {
    let mut attrs = attrs.iter().filter(|attr| attr.path() == BORSH);
    let borsh = attrs.next();
    if let Some(other_borsh) = attrs.next() {
        return Err(syn::Error::new_spanned(
            other_borsh,
            format!("multiple `{}` attributes not allowed", BORSH.name),
        ));
    }
    Ok(borsh)
}
