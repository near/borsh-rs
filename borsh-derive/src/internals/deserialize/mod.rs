use cfg_if::cfg_if;
use proc_macro2::TokenStream as TokenStream2;
use quote::quote;
use syn::{parse_quote, ExprPath, Generics, Ident, Path, Type};

use super::{
    attributes::{field, BoundType},
    generics,
};

pub mod enums;
pub mod structs;
pub mod unions;

struct GenericsOutput {
    overrides: Vec<syn::WherePredicate>,
    default_visitor: generics::FindTyParams,
    deserialize_visitor: generics::FindTyParams,
}

impl GenericsOutput {
    fn new(generics: &Generics) -> Self {
        Self {
            overrides: vec![],
            deserialize_visitor: generics::FindTyParams::new(generics),
            default_visitor: generics::FindTyParams::new(generics),
        }
    }

    fn extend<const IS_ASYNC: bool>(self, where_clause: &mut syn::WhereClause, cratename: &Path) {
        let de_trait: Path = if IS_ASYNC {
            parse_quote! { #cratename::de::BorshDeserializeAsync }
        } else {
            parse_quote! { #cratename::de::BorshDeserialize }
        };
        let default_trait: Path = parse_quote! { ::core::default::Default };
        let de_predicates =
            generics::compute_predicates(self.deserialize_visitor.process_for_bounds(), &de_trait);
        let default_predicates =
            generics::compute_predicates(self.default_visitor.process_for_bounds(), &default_trait);
        where_clause.predicates.extend(de_predicates);
        where_clause.predicates.extend(default_predicates);
        where_clause.predicates.extend(self.overrides);
    }
}

fn process_field<const IS_ASYNC: bool>(
    field: &syn::Field,
    cratename: &Path,
    body: &mut TokenStream2,
    generics: &mut GenericsOutput,
) -> syn::Result<()> {
    let parsed = field::Attributes::parse(&field.attrs)?;

    generics
        .overrides
        .extend(parsed.collect_bounds::<IS_ASYNC>(BoundType::Deserialize));
    let needs_bounds_derive = parsed.needs_bounds_derive::<IS_ASYNC>(BoundType::Deserialize);

    let field_name = field.ident.as_ref();
    let delta = if parsed.skip {
        if needs_bounds_derive {
            generics.default_visitor.visit_field(field);
        }
        field_default_output(field_name)
    } else {
        if needs_bounds_derive {
            generics.deserialize_visitor.visit_field(field);
        }
        field_output::<IS_ASYNC>(
            field_name,
            &field.ty,
            cratename,
            if IS_ASYNC {
                cfg_if! {
                    if #[cfg(feature = "async")] {
                        parsed.deserialize_with_async
                    } else {
                        None
                    }
                }
            } else {
                parsed.deserialize_with
            },
        )
    };
    body.extend(delta);
    Ok(())
}

/// function which computes derive output [proc_macro2::TokenStream]
/// of code, which deserializes single field
fn field_output<const IS_ASYNC: bool>(
    field_name: Option<&Ident>,
    field_type: &Type,
    cratename: &Path,
    deserialize_with: Option<ExprPath>,
) -> TokenStream2 {
    let default_path = || {
        let deserialize_trait = Ident::new(
            if IS_ASYNC {
                "BorshDeserializeAsync"
            } else {
                "BorshDeserialize"
            },
            proc_macro2::Span::call_site(),
        );
        parse_quote! { <#field_type as #cratename::#deserialize_trait>::deserialize_reader }
    };

    let path: ExprPath = deserialize_with.unwrap_or_else(default_path);
    let dot_await = IS_ASYNC.then(|| quote! { .await });

    if let Some(field_name) = field_name {
        quote! { #field_name: #path(reader)#dot_await?, }
    } else {
        quote! { #path(reader)#dot_await?, }
    }
}

/// function which computes derive output [proc_macro2::TokenStream]
/// of code, which deserializes single skipped field
fn field_default_output(field_name: Option<&Ident>) -> TokenStream2 {
    if let Some(field_name) = field_name {
        quote! { #field_name: ::core::default::Default::default(), }
    } else {
        quote! { ::core::default::Default::default(), }
    }
}
