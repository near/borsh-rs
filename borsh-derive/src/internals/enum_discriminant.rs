use std::{collections::HashMap, convert::TryFrom};

use proc_macro2::{Ident, TokenStream};
use quote::{quote, ToTokens};
use syn::{punctuated::Punctuated, token::Comma, Variant};

pub struct Discriminants(HashMap<Ident, TokenStream>);
impl Discriminants {
    /// Calculates the discriminant that will be assigned by the compiler.
    /// See: https://doc.rust-lang.org/reference/items/enumerations.html#assigning-discriminant-values
    pub fn new(variants: &Punctuated<Variant, Comma>) -> Self {
        let mut map = HashMap::new();
        let mut next_discriminant_if_not_specified = quote! {0};

        for variant in variants {
            let this_discriminant = variant.discriminant.clone().map_or_else(
                || next_discriminant_if_not_specified,
                |(_, e)| e.into_token_stream(),
            );

            next_discriminant_if_not_specified = quote! { #this_discriminant + 1 };
            map.insert(variant.ident.clone(), this_discriminant);
        }

        Self(map)
    }

    pub fn get(
        &self,
        variant_ident: &Ident,
        use_discriminant: bool,
        variant_idx: usize,
    ) -> syn::Result<TokenStream> {
        let variant_idx = u8::try_from(variant_idx).map_err(|err| {
            syn::Error::new(
                variant_ident.span(),
                format!("up to 256 enum variants are supported: {}", err),
            )
        })?;
        let result = if use_discriminant {
            self.0.get(variant_ident).unwrap().clone() // discriminant value
        } else {
            quote! { #variant_idx }
        };
        Ok(result)
    }
}
