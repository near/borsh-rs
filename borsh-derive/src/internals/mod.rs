pub mod attributes;
pub mod deserialize;
mod enum_discriminant;
mod generics;
#[cfg(feature = "schema")]
pub mod schema;
pub mod serialize;

pub mod cratename;

#[cfg(test)]
mod test_helpers;
