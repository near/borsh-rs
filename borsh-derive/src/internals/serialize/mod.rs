use std::convert::TryFrom;

use proc_macro2::{Span, TokenStream as TokenStream2};
use quote::quote;
use syn::{parse_quote, Expr, ExprPath, Generics, Ident, Index, Path, Type};

use super::generics;

pub mod enums;
pub mod structs;
pub mod unions;

struct GenericsOutput {
    overrides: Vec<syn::WherePredicate>,
    serialize_visitor: generics::FindTyParams,
}

impl GenericsOutput {
    fn new(generics: &Generics) -> Self {
        Self {
            overrides: vec![],
            serialize_visitor: generics::FindTyParams::new(generics),
        }
    }
    fn extend<const IS_ASYNC: bool>(self, where_clause: &mut syn::WhereClause, cratename: &Path) {
        let trait_path: Path = if IS_ASYNC {
            parse_quote! { #cratename::ser::BorshSerializeAsync }
        } else {
            parse_quote! { #cratename::ser::BorshSerialize }
        };
        let predicates =
            generics::compute_predicates(self.serialize_visitor.process_for_bounds(), &trait_path);
        where_clause.predicates.extend(predicates);
        where_clause.predicates.extend(self.overrides);
    }
}

pub enum FieldId {
    Struct(Ident),
    StructUnnamed(Index),
    Enum(Ident),
    EnumUnnamed(Index),
}

impl FieldId {
    fn index(field_idx: usize) -> syn::Result<Index> {
        let index = u32::try_from(field_idx).map_err(|err| {
            syn::Error::new(
                Span::call_site(),
                format!("up to 2^32 fields are supported {}", err),
            )
        })?;
        Ok(Index {
            index,
            span: Span::call_site(),
        })
    }
    pub fn new_struct_unnamed(field_idx: usize) -> syn::Result<Self> {
        let index = Self::index(field_idx)?;
        let result = Self::StructUnnamed(index);
        Ok(result)
    }
    pub fn new_enum_unnamed(field_idx: usize) -> syn::Result<Self> {
        let index = Self::index(field_idx)?;
        let result = Self::EnumUnnamed(index);
        Ok(result)
    }
}

impl FieldId {
    fn serialize_arg(&self) -> Expr {
        match self {
            Self::Struct(name) => parse_quote! { &self.#name },
            Self::StructUnnamed(index) => parse_quote! { &self.#index },
            Self::Enum(name) => parse_quote! { #name },
            Self::EnumUnnamed(ind) => {
                let field = Ident::new(&format!("id{}", ind.index), Span::mixed_site());
                parse_quote! { #field }
            }
        }
    }

    /// function which computes derive output [proc_macro2::TokenStream]
    /// of code, which serializes single field
    pub fn serialize_output<const IS_ASYNC: bool>(
        &self,
        field_type: &Type,
        cratename: &Path,
        serialize_with: Option<ExprPath>,
    ) -> TokenStream2 {
        let arg: Expr = self.serialize_arg();
        let dot_await = IS_ASYNC.then(|| quote! { .await });
        if let Some(func) = serialize_with {
            quote! { #func(#arg, writer)#dot_await?; }
        } else {
            let serialize_trait = Ident::new(
                if IS_ASYNC {
                    "BorshSerializeAsync"
                } else {
                    "BorshSerialize"
                },
                Span::call_site(),
            );
            quote! { <#field_type as #cratename::#serialize_trait>::serialize(#arg, writer)#dot_await?; }
        }
    }

    pub fn enum_variant_header(&self, skipped: bool) -> Option<TokenStream2> {
        match self {
            Self::Struct(..) | Self::StructUnnamed(..) => unreachable!("no variant header"),
            Self::Enum(name) => (!skipped).then_some(quote! { #name, }),
            Self::EnumUnnamed(index) => {
                let field_ident = Ident::new(
                    &if skipped {
                        format!("_id{}", index.index)
                    } else {
                        format!("id{}", index.index)
                    },
                    Span::mixed_site(),
                );
                Some(quote! { #field_ident, })
            }
        }
    }
}
