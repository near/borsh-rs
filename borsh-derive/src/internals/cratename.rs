use proc_macro2::Span;
use proc_macro_crate::{crate_name, FoundCrate};
use syn::{Attribute, Error, Ident, Path};

use super::attributes::item;

pub(crate) const BORSH: &str = "borsh";

pub(crate) fn get(attrs: &[Attribute]) -> Result<Path, Error> {
    let path = item::get_crate(attrs)?;
    match path {
        Some(path) => Ok(path),
        None => {
            let ident = get_from_cargo();
            Ok(ident.into())
        }
    }
}

pub(crate) fn get_from_cargo() -> Ident {
    let name = &crate_name(BORSH).unwrap();
    let name = match name {
        FoundCrate::Itself => BORSH,
        FoundCrate::Name(name) => name.as_str(),
    };
    Ident::new(name, Span::call_site())
}
