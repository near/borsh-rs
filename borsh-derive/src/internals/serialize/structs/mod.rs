use cfg_if::cfg_if;
use proc_macro2::{Ident, Span, TokenStream as TokenStream2};
use quote::quote;
use syn::{Fields, ItemStruct, Lifetime, Path, Token};

use crate::internals::{
    attributes::{field, BoundType},
    generics, serialize,
};

pub fn process<const IS_ASYNC: bool>(
    input: ItemStruct,
    cratename: Path,
) -> syn::Result<TokenStream2> {
    let name = &input.ident;
    let generics = generics::without_defaults(input.generics);
    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
    let mut where_clause = generics::default_where(where_clause);
    let mut body = TokenStream2::new();
    let mut generics_output = serialize::GenericsOutput::new(&generics);
    match &input.fields {
        Fields::Named(fields) => {
            for field in &fields.named {
                let field_id = serialize::FieldId::Struct(field.ident.clone().unwrap());

                process_field::<IS_ASYNC>(
                    field,
                    field_id,
                    &cratename,
                    &mut generics_output,
                    &mut body,
                )?;
            }
        }
        Fields::Unnamed(fields) => {
            for (field_idx, field) in fields.unnamed.iter().enumerate() {
                let field_id = serialize::FieldId::new_struct_unnamed(field_idx)?;

                process_field::<IS_ASYNC>(
                    field,
                    field_id,
                    &cratename,
                    &mut generics_output,
                    &mut body,
                )?;
            }
        }
        Fields::Unit => {}
    }
    generics_output.extend::<IS_ASYNC>(&mut where_clause, &cratename);

    let serialize_trait = Ident::new(
        if IS_ASYNC {
            "BorshSerializeAsync"
        } else {
            "BorshSerialize"
        },
        Span::call_site(),
    );
    let writer_trait_path = if IS_ASYNC {
        quote! { async_io::AsyncWrite }
    } else {
        quote! { io::Write }
    };
    let r#async = IS_ASYNC.then(|| Token![async](Span::call_site()));
    let lifetime = IS_ASYNC.then(|| Lifetime::new("'async_variant", Span::call_site()));
    let lt_comma = IS_ASYNC.then(|| Token![,](Span::call_site()));

    Ok(quote! {
        impl #impl_generics #cratename::ser::#serialize_trait for #name #ty_generics #where_clause {
            #r#async fn serialize<#lifetime #lt_comma __W: #cratename::#writer_trait_path>(
                &#lifetime self,
                writer: &#lifetime mut __W,
            ) -> ::core::result::Result<(), #cratename::io::Error> {
                #body
                ::core::result::Result::Ok(())
            }
        }
    })
}

fn process_field<const IS_ASYNC: bool>(
    field: &syn::Field,
    field_id: serialize::FieldId,
    cratename: &Path,
    generics: &mut serialize::GenericsOutput,
    body: &mut TokenStream2,
) -> syn::Result<()> {
    let parsed = field::Attributes::parse(&field.attrs)?;

    let needs_bounds_derive = parsed.needs_bounds_derive::<IS_ASYNC>(BoundType::Serialize);
    generics
        .overrides
        .extend(parsed.collect_bounds::<IS_ASYNC>(BoundType::Serialize));

    if !parsed.skip {
        let delta = field_id.serialize_output::<IS_ASYNC>(
            &field.ty,
            cratename,
            if IS_ASYNC {
                cfg_if! {
                    if #[cfg(feature = "async")] {
                        parsed.serialize_with_async
                    } else {
                        None
                    }
                }
            } else {
                parsed.serialize_with
            },
        );
        body.extend(delta);

        if needs_bounds_derive {
            generics.serialize_visitor.visit_field(field);
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use syn::parse_quote;

    use super::*;
    use crate::internals::test_helpers::{
        default_cratename, local_insta_assert_debug_snapshot, local_insta_assert_snapshot,
        pretty_print_syn_str,
    };

    #[test]
    fn simple_struct() {
        let item_struct: ItemStruct = parse_quote! {
            struct A {
                x: u64,
                y: String,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn simple_struct_with_custom_crate() {
        let item_struct: ItemStruct = parse_quote! {
            struct A {
                x: u64,
                y: String,
            }
        };

        let crate_: Path = parse_quote! { reexporter::borsh };

        let actual = process::<false>(item_struct.clone(), crate_.clone()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, crate_).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn simple_generics() {
        let item_struct: ItemStruct = parse_quote! {
            struct A<K, V> {
                x: HashMap<K, V>,
                y: String,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn simple_generic_tuple_struct() {
        let item_struct: ItemStruct = parse_quote! {
            struct TupleA<T>(T, u32);
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn bound_generics() {
        let item_struct: ItemStruct = parse_quote! {
            struct A<K: Key, V> where V: Value {
                x: HashMap<K, V>,
                y: String,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn recursive_struct() {
        let item_struct: ItemStruct = parse_quote! {
            struct CRecC {
                a: String,
                b: HashMap<String, CRecC>,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn generic_tuple_struct_borsh_skip1() {
        let item_struct: ItemStruct = parse_quote! {
            struct G<K, V, U> (
                #[borsh(skip)]
                HashMap<K, V>,
                U,
            );
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn generic_tuple_struct_borsh_skip2() {
        let item_struct: ItemStruct = parse_quote! {
            struct G<K, V, U> (
                HashMap<K, V>,
                #[borsh(skip)]
                U,
            );
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn generic_named_fields_struct_borsh_skip() {
        let item_struct: ItemStruct = parse_quote! {
            struct G<K, V, U> {
                #[borsh(skip)]
                x: HashMap<K, V>,
                y: U,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn generic_associated_type() {
        let item_struct: ItemStruct = parse_quote! {
            struct Parametrized<T, V>
            where
                T: TraitName,
            {
                field: T::Associated,
                another: V,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn generic_serialize_bound() {
        let item_struct: ItemStruct = parse_quote! {
            struct C<T: Debug, U> {
                a: String,
                #[borsh(bound(serialize =
                    "T: borsh::ser::BorshSerialize + PartialOrd,
                     U: borsh::ser::BorshSerialize"
                ))]
                b: HashMap<T, U>,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    #[cfg(feature = "async")]
    fn generic_serialize_async_bound() {
        let item_struct: ItemStruct = parse_quote! {
            struct C<T: Debug, U> {
                a: String,
                #[borsh(async_bound(serialize =
                    "T: borsh::ser::BorshSerializeAsync + PartialOrd,
                     U: borsh::ser::BorshSerializeAsync"
                ))]
                b: HashMap<T, U>,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        let actual = process::<true>(item_struct, default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
    }

    #[test]
    fn override_generic_associated_type_wrong_derive() {
        let item_struct: ItemStruct = parse_quote! {
            struct Parametrized<T, V> where T: TraitName {
                #[borsh(bound(serialize =
                    "<T as TraitName>::Associated: borsh::ser::BorshSerialize"
                ))]
                field: <T as TraitName>::Associated,
                another: V,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    #[cfg(feature = "async")]
    fn async_override_generic_associated_type_wrong_derive() {
        let item_struct: ItemStruct = parse_quote! {
            struct Parametrized<T, V> where T: TraitName {
                #[borsh(async_bound(serialize =
                    "<T as TraitName>::Associated: borsh::ser::BorshSerializeAsync"
                ))]
                field: <T as TraitName>::Associated,
                another: V,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        let actual = process::<true>(item_struct, default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
    }

    #[test]
    fn check_serialize_with_attr() {
        let item_struct: ItemStruct = parse_quote! {
            struct A<K: Ord, V> {
                #[borsh(serialize_with = "third_party_impl::serialize_third_party")]
                x: ThirdParty<K, V>,
                y: u64,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_struct, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    #[cfg(feature = "async")]
    fn check_serialize_with_async_attr() {
        let item_struct: ItemStruct = parse_quote! {
            struct A<K: Ord, V> {
                #[borsh(serialize_with_async = "third_party_impl::serialize_third_party")]
                x: ThirdParty<K, V>,
                y: u64,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        let actual = process::<true>(item_struct, default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
    }

    #[test]
    #[cfg(not(feature = "async"))]
    fn check_serialize_with_skip_conflict() {
        let item_struct: ItemStruct = parse_quote! {
            struct A<K: Ord, V> {
                #[borsh(skip, serialize_with = "third_party_impl::serialize_third_party")]
                x: ThirdParty<K, V>,
                y: u64,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename());

        let err = match actual {
            Ok(..) => unreachable!("expecting error here"),
            Err(err) => err,
        };
        local_insta_assert_debug_snapshot!(err);
    }

    #[test]
    #[cfg(feature = "async")]
    fn check_serialize_with_skip_conflict_feature_async() {
        let item_struct: ItemStruct = parse_quote! {
            struct A<K: Ord, V> {
                #[borsh(skip, serialize_with = "third_party_impl::serialize_third_party")]
                x: ThirdParty<K, V>,
                y: u64,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename());

        let err = match actual {
            Ok(..) => unreachable!("expecting error here"),
            Err(err) => err,
        };
        local_insta_assert_debug_snapshot!(err);

        let actual = process::<true>(item_struct, default_cratename());

        let err = match actual {
            Ok(..) => unreachable!("expecting error here"),
            Err(err) => err,
        };
        local_insta_assert_debug_snapshot!(err);
    }

    #[test]
    #[cfg(feature = "async")]
    fn check_serialize_with_async_skip_conflict() {
        let item_struct: ItemStruct = parse_quote! {
            struct A<K: Ord, V> {
                #[borsh(skip, serialize_with_async = "third_party_impl::serialize_third_party")]
                x: ThirdParty<K, V>,
                y: u64,
            }
        };

        let actual = process::<false>(item_struct.clone(), default_cratename());

        let err = match actual {
            Ok(..) => unreachable!("expecting error here"),
            Err(err) => err,
        };
        local_insta_assert_debug_snapshot!(err);

        let actual = process::<true>(item_struct, default_cratename());

        let err = match actual {
            Ok(..) => unreachable!("expecting error here"),
            Err(err) => err,
        };
        local_insta_assert_debug_snapshot!(err);
    }
}
