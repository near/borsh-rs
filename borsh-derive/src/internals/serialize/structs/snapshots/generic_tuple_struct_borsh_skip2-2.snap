---
source: borsh-derive/src/internals/serialize/structs/mod.rs
expression: pretty_print_syn_str(actual).unwrap()
---
impl<K, V, U> borsh::ser::BorshSerializeAsync for G<K, V, U>
where
    K: borsh::ser::BorshSerializeAsync,
    V: borsh::ser::BorshSerializeAsync,
{
    async fn serialize<'async_variant, __W: borsh::async_io::AsyncWrite>(
        &'async_variant self,
        writer: &'async_variant mut __W,
    ) -> ::core::result::Result<(), borsh::io::Error> {
        <HashMap<K, V> as borsh::BorshSerializeAsync>::serialize(&self.0, writer).await?;
        ::core::result::Result::Ok(())
    }
}
